File: src/fable-compiler-js/fable-library/Range.js
import { compare } from "./Util.js";
import { unfold, delay } from "./Seq.js";
import { toUInt64, toInt64, op_Addition, fromZero } from "./BigInt.js";
import { op_Addition as op_Addition_1, fromParts } from "./Decimal.js";
export function makeRangeStepFunction(step, stop, zero, add) {
    const stepComparedWithZero = compare(step, zero) | 0;
    if (stepComparedWithZero === 0) {
        throw new Error("The step of a range cannot be zero");
    }
    const stepGreaterThanZero = stepComparedWithZero > 0;
    return (x) => {
        const comparedWithLast = compare(x, stop) | 0;
        return ((stepGreaterThanZero && (comparedWithLast <= 0)) ? true : (!stepGreaterThanZero && (comparedWithLast >= 0))) ? [x, add(x, step)] : void 0;
    };
}
export function integralRangeStep(start, step, stop, zero, add) {
    const stepFn = makeRangeStepFunction(step, stop, zero, add);
    return delay(() => unfold(stepFn, start));
}
export function rangeBigInt(start, step, stop) {
    return integralRangeStep(start, step, stop, fromZero(), op_Addition);
}
export function rangeDecimal(start, step, stop) {
    return integralRangeStep(start, step, stop, fromParts(0, 0, 0, false, 0), op_Addition_1);
}
export function rangeDouble(start, step, stop) {
    return integralRangeStep(start, step, stop, 0, (x, y) => (x + y));
}
export function rangeInt64(start, step, stop) {
    return integralRangeStep(start, step, stop, 0n, (x, y) => toInt64(op_Addition(x, y)));
}
export function rangeUInt64(start, step, stop) {
    return integralRangeStep(start, step, stop, 0n, (x, y) => toUInt64(op_Addition(x, y)));
}
export function rangeChar(start, stop) {
    const intStop = stop.charCodeAt(0) | 0;
    return delay(() => unfold((c) => {
        if (c <= intStop) {
            return [String.fromCharCode(c), c + 1];
        }
        else {
            return void 0;
        }
    }, start.charCodeAt(0)));
}

File: src/fable-library-dart/Range.fs
module FSharp.Core.OperatorIntrinsics

let makeRangeStepFunction<'T>
    (step: 'T)
    (stop: 'T)
    (zero: 'T)
    (add: 'T -> 'T -> 'T)
    (compare: 'T -> 'T -> int)
    =
    let stepComparedWithZero = compare step zero

    if stepComparedWithZero = 0 then
        failwith "The step of a range cannot be zero"

    let stepGreaterThanZero = stepComparedWithZero > 0

    fun x ->
        let comparedWithLast = compare x stop

        if
            (stepGreaterThanZero && comparedWithLast <= 0)
            || (not stepGreaterThanZero && comparedWithLast >= 0)
        then
            Some(x, add x step)
        else
            None

let integralRangeStep<'T>
    (start: 'T)
    (step: 'T)
    (stop: 'T)
    (zero: 'T)
    (add: 'T -> 'T -> 'T)
    (compare: 'T -> 'T -> int)
    =
    let stepFn = makeRangeStepFunction step stop zero add compare
    Seq.delay (fun () -> Seq.unfold stepFn start)

let rangeDouble start step stop =
    integralRangeStep start step stop 0.0 (+) compare

let rangeInt start step stop =
    integralRangeStep start step stop 0 (+) compare
//let rangeBigInt start step stop = integralRangeStep start step stop 0I (+) compare
//let rangeDecimal start step stop = integralRangeStep start step stop 0m (+) compare

let rangeChar (start: char) (stop: char) =
    let intStop = int stop

    let stepFn c =
        if c <= intStop then
            Some(char c, c + 1)
        else
            None

    Seq.delay (fun () -> Seq.unfold stepFn (int start))
